### PR TITLE
Jburdass profile edit page

### PR DESCRIPTION
**Pull request by Benjipictor**

### CODE REVIEW

File: src/components/saveChangesModal/index.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import './style.css'
+import Button from '../button'
+import Toast from '../toast'
+import { patch } from "../../service/apiClient";
+
+const SaveChangesModal = ({ formState, id, loggedInUserInfo }) => {
+
+    const { closeModal } = useModal()
+    const navigate = useNavigate()
+    console.log("modal is working")
```
Comment (vherus) [inline src/components/saveChangesModal/index.js]: Let's remove any debug logs

File: src/components/saveChangesModal/index.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    const token = localStorage.getItem("token");
```
Comment (vherus) [inline src/components/saveChangesModal/index.js]: Use the `useAuth` hook to get the token, `const { token } = useAuth()`

File: src/components/saveChangesModal/index.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    const handleSave = () => {
+        handleSubmit()
+        closeModal()
+        navigate(`/profile/${id}`)
+        return (
+            <Toast text="profile saved!" linkText="Edit" linkTo="/profile/edit" />
+        )
+    }
+    const handleSubmit = () => {
+        console.log("form submitted")
```
Comment (vherus) [inline src/components/saveChangesModal/index.js]: Debug log

File: src/components/saveChangesModal/index.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        // console.log("here is the edited profilejson > ", editedProfileJSON)
+
+        // const options = {
+        //     method: "PATCH",
+        //     body: editedProfileJSON,
+        //     headers: {
+        //         "Content-Type": "application/json",
+        //         Authorization: "Bearer " + token
+        //     }
+        // }
```
Comment (vherus) [inline src/components/saveChangesModal/index.js]: Please remove this dead code

File: src/components/saveChangesModal/index.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        const formDataPATCH = async () => {
+            const endpoint = `users/${id}`
+
+            // editedProfile.cohortId = editedProfile.cohort_id
+            // const data = JSON.stringify(editedProfile)
```
Comment (vherus) [inline src/components/saveChangesModal/index.js]: Dead code

File: src/components/saveChangesModal/index.js [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+                    password: editedProfile.password,
+                    firstName: editedProfile.firstName,
+                    lastName: editedProfile.lastName,
+                    biography: editedProfile.biography,
+                    githubUrl: editedProfile.githubUrl,
+                    role: editedProfile.role,
+                    cohortId: editedProfile.cohortId
+                }
+                console.log("here is the data", data)
+                console.log("this is the edited profile", editedProfile)
```
Comment (vherus) [inline src/components/saveChangesModal/index.js]: Debug logs

File: src/components/saveChangesModal/index.js [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+                patch(endpoint, editedProfile)
+                    .catch((error) => { console.log(error) })
```
Comment (vherus) [inline src/components/saveChangesModal/index.js]: Instead of logging an error to the console, this should display a useful message to the user

File: src/components/saveChangesModal/index.js [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+                const data = {
+                    email: editedProfile.email,
+                    password: editedProfile.password,
+                    firstName: editedProfile.firstName,
+                    lastName: editedProfile.lastName,
+                    biography: editedProfile.biography,
+                    githubUrl: editedProfile.githubUrl,
+                    role: editedProfile.role,
+                    cohortId: editedProfile.cohortId
+                }
```
Comment (vherus) [inline src/components/saveChangesModal/index.js]: You're creating this `data` variable within the scope of these if statements, so it won't be accessible outside of the if statements. Here's a better approach:

```js
const data = {
    email: editedProfile.email,
    password: editedProfile.password,
    firstName: editedProfile.firstName,
    lastName: editedProfile.lastName,
    biography: editedProfile.biography,
    githubUrl: editedProfile.githubUrl
}

if (loggedInUserInfo.role === "TEACHER") {
    data.role = editedProfile.role
    data.cohortId = editedProfile.cohortId
}
```

File: src/pages/editProfile/index.js [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+  const ProfileImg = () => {
+    if (profile.profileImageUrl === "") {
+      return <ProfileCircle
+        initials={`${profile.firstName[0]} ${profile.lastName[0]}`}
+      />
+    }
+    else { return (<img className="profile-icon" src={profile.profileImageUrl} alt="profile Image"></img>); }
+
+  };
```
Comment (vherus) [inline src/pages/editProfile/index.js]: Let's move this to its own component file

File: src/test.js [hunk 10]
```diff
@@ -0,0 +1,8 @@
+const Test = () => {
+ console.log("test")
+ return (
+    <p>Test</p>
+ )
+}
+
+export default Test
```
Comment (vherus) [inline src/test.js]: What's the purpose of this component? If it has none, please remove it

**Review by vherus — CHANGES_REQUESTED** [review]
Nice work, there are some changes to make here

File: src/components/saveChangesModal/index.js [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+            }
+            if (loggedInUserInfo.role === "TEACHER") {
+                data.role = editedProfile.role
+                data.cohortId = editedProfile.cohortId
+
+                patch(endpoint, editedProfile)
+                    .catch((error) => { setIsError(true) })
+
+            }
+            formDataPATCH()
```
Comment (vherus) [inline src/components/saveChangesModal/index.js]: This `formDataPATCH` call is inside the `formDataPATCH` function, you'll end up with an endless recursive function calling itself until the app crashes if this runs

**Review by vherus — DISMISSED** [review]

**Comment by Logi13** [comment]
image
There is somewhere a read-only component that something tries to write to it.

**Comment by Logi13** [comment]
There is no patch sent to the server

File: src/pages/editProfile/index.js [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+  }
+  const profileData = async () => {
+    const res = await get(`users/${id}`);
+    res.data.user ? setFormState(res.data.user) : setIsError(true);
+    res.data.user ? setProfile(res.data.user) : setIsError(true)
+  };
+
+  const controlPagePermission = () => {
+    if (loggedInUserInfo.role === "STUDENT" && loggedInUserInfo.id !== Number(id)) {
+      navigate(-1)
```
Comment (Logi13) [inline src/pages/editProfile/index.js]: just redirect back to the current user's profile

File: src/pages/editProfile/index.js [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+    }
+    if (loggedInUserInfo.role === "STUDENT") {
+      setReadOnly(true)
+    }
+    if (loggedInUserInfo.id === Number(id)) {
+      setPasswordPermission(false)
```
Comment (Logi13) [inline src/pages/editProfile/index.js]: I would expect if I am the user to be able to change my password to set passwword permissions to true

File: src/components/saveChangesModal/index.js [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+
+    const handleCancel = () => {
+        closeModal()
+        navigate(`/profile/${id}/edit`)
+    }
+
+    const handleSave = () => {
+        handleSubmit()
+        if(isError === true){
+
```
Comment (Logi13) [inline src/components/saveChangesModal/index.js]: just a console.log saying that there is some error happening

File: src/components/saveChangesModal/index.js [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+                firstName: editedProfile.firstName,
+                lastName: editedProfile.lastName,
+                biography: editedProfile.biography,
+                githubUrl: editedProfile.githubUrl
+            }
+            if (loggedInUserInfo.role === "TEACHER") {
+                data.role = editedProfile.role
+                data.cohortId = editedProfile.cohortId
+
+                patch(endpoint, editedProfile)
```
Comment (Logi13) [inline src/components/saveChangesModal/index.js]: patch will not work if the user is a student!

File: src/pages/editProfile/index.js [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+
+
+const EditProfile = () => {
+  const [profile, setProfile] = useState(emptyProfile);
+  const [formState, setFormState] = useState(profile);
+  const [readOnly, setReadOnly] = useState(false)
+  const [passwordPermission, setPasswordPermission] = useState(true)
+  const [isError, setIsError] = useState(false);
+  const navigate = useNavigate()
+  const { id } = useParams()
```
Comment (Logi13) [inline src/pages/editProfile/index.js:26]: add a userId with a setter state

on the useEffect check if the id is different than the userid and update the userId

**Review by Logi13 — DISMISSED** [review]

**Comment by Logi13** [comment]
Good work by the way, except those small issue you should be fine :)

File: src/pages/editProfile/index.js [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+      setReadOnly(true)
+    }
+    if (loggedInUserInfo.id === Number(id)) {
+      setPasswordPermission(false)
+    }
+  }
+  useEffect(() => {
+    profileData()
+  }, [id])
+  useEffect(() => {
```
Comment (Logi13) [inline src/pages/editProfile/index.js]: merge the 2 useEffects

File: src/pages/editProfile/index.js [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+                    label="Specialism*"
+                    name="specialism"
+                    value={formState.specialism}
+                    onChange={handleChange}
+                    readOnly={readOnly}
+                  />
+                  <TextInput
+                    label="Cohort*"
+                    name="cohortId"
+                    value={formState.cohortId}
```
Comment (julesnuggy) [inline src/pages/editProfile/index.js]: The BE response returns this as `cohort_id`

File: src/pages/editProfile/index.js [hunk 19]
```diff
@@ -0,0 +1,244 @@
...
+          <Card>
+            <div className="profile-header">
+              <ProfileCircle
+                initials={`${profile.firstName[0]} ${profile.lastName[0]}`}
+              />
+              <div className="profile-header-text">
+                <h4>
+                  {profile.firstName} {profile.lastName}
+                </h4>
+                <p>{profile.specialism}this is the specialism</p>
```
Comment (julesnuggy) [inline src/pages/editProfile/index.js]: We want to get rid of the placeholder text.

File: src/pages/editProfile/index.js [hunk 20]
```diff
@@ -0,0 +1,244 @@
...
+              </div>
+            </div>
+
+            <form onSubmit={showModal}>
+              <section className="basicInfoSection text-input-containers">
+                <h2>Basic Info</h2>
+                <div className="headshot-container">
+                  <ProfileImg profile={profile} />
+                  <p>Add headshot</p>
```
Comment (julesnuggy) [inline src/pages/editProfile/index.js:106]: Do we need this "Add headshot" element here?

File: src/pages/editProfile/index.js [hunk 21]
```diff
@@ -0,0 +1,244 @@
...
+import ProfileImg from "../../components/profileImage/ProfileImage";
+
+
+const EditProfile = () => {
+  const [profile, setProfile] = useState(emptyProfile);
+  const [formState, setFormState] = useState(emptyProfile);
+  const [readOnly, setReadOnly] = useState(false)
+  const [passwordPermission, setPasswordPermission] = useState(false)
+  const [isError, setIsError] = useState(false);
+  const [noPassword, setNopassword] = useState(false)
```
Comment (julesnuggy) [inline src/pages/editProfile/index.js]: Currently, the default is `noPassword = false` which means there is a password, which doesn't feel like the right default.

Rather than dealing with double-negatives, it is usually easier to have state (and variables in general) be phrased in the positive, so this could instead be `hasUserEnteredPassword` with a default of `false`. Remember to check that the logic where this is used still works/makes sense.

File: src/pages/editProfile/index.js [hunk 22]
```diff
@@ -0,0 +1,244 @@
...
+                />
+                <TextInput
+                  label="Password*"
+                  name="password"
+                  value={formState.password}
+                  type="password"
+                  onChange={handleChange}
+                  permission={passwordPermission}
+                />
+                {noPassword && <ErrorMessage message={"Pleas add a password to submit changes"} />}
```
Comment (julesnuggy) [inline src/pages/editProfile/index.js]: Typo with "Pleas"

File: src/pages/editProfile/index.js [hunk 23]
```diff
@@ -0,0 +1,244 @@
...
+  const { id } = useParams()
+  const [userId, setUserId] = useState(id)
+
+  const { openModal, setModal } = useModal();
+  const { loggedInUserInfo } = useAuth();
+
+  const showModal = (event) => {
+    event.preventDefault()
+    if(!formState.password){
+      setNopassword(true)
```
Comment (julesnuggy) [inline src/pages/editProfile/index.js]: You'll need to set this to `false` in the else block, otherwise the warning message will not disappear.

File: src/components/saveChangesModal/index.js [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+import { patch } from "../../service/apiClient";
+import ErrorMessage from "../errorMessage";
+
+const SaveChangesModal = ({ formState, id, loggedInUserInfo }) => {
+    const [isError, setIsError] = useState(false)
+    const { closeModal } = useModal()
+    const navigate = useNavigate()
+    const handleDontSave = () => {       
+            closeModal()
+            navigate(`/profile/${id}`)
```
Comment (julesnuggy) [inline src/components/saveChangesModal/index.js]: Fix the indentation please.

File: src/components/saveChangesModal/index.js [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+        const editedProfile = formState
+        const formDataPATCH = async () => {         
+            const endpoint = `users/${id}`
+            const data = {
+                email: editedProfile.email,
+                firstName: editedProfile.firstName,
+                lastName: editedProfile.lastName,
+                biography: editedProfile.biography,
+                githubUrl: editedProfile.githubUrl
+                }
```
Comment (julesnuggy) [inline src/components/saveChangesModal/index.js]: Fix indentation please.

File: src/components/saveChangesModal/index.js [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+                if(editedProfile.password !== undefined){
```
Comment (julesnuggy) [inline src/components/saveChangesModal/index.js]: This can be `!editedProfile.password`

File: src/components/saveChangesModal/index.js [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+                    data.password = editedProfile.password
+                }
+            if (loggedInUserInfo.role === "TEACHER") {
+                data.role = editedProfile.role
+                data.cohortId = editedProfile.cohortId
+              }
+            const res = await patch(endpoint, data)
+            console.log(res)
```
Comment (julesnuggy) [inline src/components/saveChangesModal/index.js]: Console log can be removed.

Comment (julesnuggy) [reply src/components/saveChangesModal/index.js]: And `res` variable is not needed anymore either.

File: src/components/saveChangesModal/index.js [hunk 28]
```diff
@@ -0,0 +1,89 @@
...
+            navigate(`/profile/${id}`)
+        }
+        if(isError === true){
+            console.log("there is an error")
+        }
+      
+    }
+    const handleSubmit = async () => {
+        const editedProfile = formState
+        const formDataPATCH = async () => {
```
Comment (julesnuggy) [inline src/components/saveChangesModal/index.js]: This function is not needed since you have defined `handleSubmit` as `async` already.

File: src/components/saveChangesModal/index.js [hunk 29]
```diff
@@ -0,0 +1,89 @@
...
+            <section className="save-changes-details">
+                <h4>Save changes to profile?</h4>
+                <p>Do you want to save these changes to your profile?</p>
+                {isError === true && <ErrorMessage message={"Cannot save changes"} />}
+            </section>
+
+            <section className="save-changes-actions">
+                <Button
+                    onClick={handleDontSave}
+                    text='dont save'
```
Comment (julesnuggy) [inline src/components/saveChangesModal/index.js]: Missing the `'` in `don't`

File: src/components/saveChangesModal/index.js [hunk 30]
```diff
@@ -0,0 +1,89 @@
...
+            closeModal()
+            navigate(`/profile/${id}`)
+        }
+        if(isError === true){
+            console.log("there is an error")
+        }
+      
+    }
+    const handleSubmit = async () => {
+        const editedProfile = formState
```
Comment (julesnuggy) [inline src/components/saveChangesModal/index.js]: Unnecessary variable - can just use `formState` directly.

File: src/components/saveChangesModal/style.css [hunk 31]
```diff
@@ -0,0 +1,25 @@
+/* This is taken from editPostModal */
+.save-changes-details {
+    display: inline;
+}
+
+.save-changes-details {
+
+}
+
```
Comment (julesnuggy) [inline src/components/saveChangesModal/style.css]: Empty style can be removed

File: src/pages/dashboard/index.js [hunk 32]
```diff
@@ -6,6 +6,7 @@ import CreatePostModal from "../../components/createPostModal";
 import TextInput from "../../components/form/textInput";
 import Posts from "../../components/posts";
 import useModal from "../../hooks/useModal";
+
```
Comment (julesnuggy) [inline src/pages/dashboard/index.js:9]: Unnecessary change.

**Review by julesnuggy — DISMISSED** [review]
Getting there! A few things to tighten up on for code quality and functionality.

File: src/pages/editProfile/index.js (hunk 20 above)
Comment (Benjipictor) [reply src/pages/editProfile/index.js:106]: its in the wire frame can remove it if needed

File: src/components/saveChangesModal/index.js (hunk 26 above)
Comment (Logi13) [reply src/components/saveChangesModal/index.js]: I think there is a typo in the above comment. The statement should be `editedProfile.password` to match the if in the code

File: src/components/form/textInput/index.js [hunk 33]
```diff
@@ -18,9 +19,11 @@ const TextInput = ({ value, onChange, name, label, icon, type = "text" }) => {
 				/>
 				{showpassword && (
 					<input
+						readOnly={permission}
 						type="text"
 						name={name}
 						value={input}
+
```
Comment (chris2733) [inline src/components/form/textInput/index.js]: Unnecessary change, random empty line?

File: src/pages/editProfile/style.css [hunk 34]
```diff
@@ -0,0 +1,79 @@
+.editContainer {
+    grid-column: span 2;
+    margin: 15px;
+}
+
+form {
```
Comment (chris2733) [inline src/pages/editProfile/style.css]: Make this more specific, otherwise it will affect every form element in the project

File: src/pages/editProfile/style.css [hunk 35]
```diff
@@ -0,0 +1,79 @@
...
+
+.bioSection {
+    grid-area: "bio";
+}
+
+.profile-header {
+    grid-area: "profile-header";
+    display: grid;
+    grid-template-columns: 50px 1fr;
+
```
Comment (chris2733) [inline src/pages/editProfile/style.css]: Tidy up these new lines, format this css doc

File: src/pages/editProfile/index.js [hunk 36]
```diff
@@ -0,0 +1,244 @@
...
+                <textarea label="Bio" name="biography" value={formState.biography} rows={20} maxLength="300" onChange={handleChange} />
+                <label htmlFor="bio" >{formState.biography.length}/300</label>
+              </section>
+              <section className="footer">
+                <p>Required*</p>
+                <Button text={"Cancel"} onClick={cancelChanges} classes="offwhite width-full" />
+                <Button text={"save"} type={"submit"} classes="blue width-full" />
+              </section>
+            </form>
+
```
Comment (chris2733) [inline src/pages/editProfile/index.js]: Format your documents, avoid these random empty lines

File: src/components/saveChangesModal/index.js [hunk 37]
```diff
@@ -0,0 +1,86 @@
...
+            
+            if (formState.password) {
+                console.log("2. adding password")
+                data.password = formState.password
+            }
+            if (loggedInUserInfo.role === "TEACHER") {
+                data.role = formState.role
+                data.cohort_id = formState.cohort_id
+            }
+            console.log("1.handle submit, this is data",data)
```
Comment (chris2733) [inline src/components/saveChangesModal/index.js]: Remove these unnecessary console logs from debugging

File: src/components/saveChangesModal/index.js [hunk 38]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    const handleSave = async () => {
+        await handleSubmit()
+        if (isError === false) {
+            closeModal()
+            navigate(`/profile/${id}`)
+        }
+        if (isError === true) {
+            console.log("there is an error")
```
Comment (chris2733) [inline src/components/saveChangesModal/index.js]: Put a more specific error message in here, something like 'Error saving profile' etc

File: src/components/profileImage/ProfileImage.js [hunk 39]
```diff
@@ -0,0 +1,13 @@
+import ProfileCircle from "../../components/profileCircle";
+
+const ProfileImg = ({ profile }) => {
+  if (profile.profileImageUrl === "") {
+    return <ProfileCircle
+      initials={`${profile.firstName[0]} ${profile.lastName[0]}`}
+    />
+  }
+  else { return (<img className="profile-icon" src={profile.profileImageUrl} alt="profile Image"></img>); }
```
Comment (chris2733) [inline src/components/profileImage/ProfileImage.js]: Not sure if this was auto-formatted, but format this document, this line looks wierd

**Review by chris2733 — CHANGES_REQUESTED** [review]

**Review by chris2733 — APPROVED** [review]

**Review by Auenc — APPROVED** [review]
